File: solutions/cpp/grains/1/grains.cpp
#include "grains.h"

#include <cmath>

namespace grains {

unsigned long long square(const int n)
{
    return std::pow(2, n - 1);
}

unsigned long long total()
{
    return (unsigned long long) std::pow(2, 64) - 1;
}

}  // namespace grains

File: solutions/zig/grains/1/grains.zig
const std = @import("std");
const pow = std.math.pow;

pub const ChessboardError = error{IndexOutOfBounds};

pub fn square(index: usize) ChessboardError!u64 {
    if (index == 0 or index > 64)
        return ChessboardError.IndexOutOfBounds;

    return pow(u64, 2, index - 1);
}

pub fn total() u64 {
    var sum: u64 = 0;
    for (0..64) |i| {
        sum += square(i + 1) catch unreachable;
    }
    return sum;
}
